source/common/stats/symbol_table_impl.cc

```diff
@@ -126,69 +126,109 @@ SymbolVec SymbolTableImpl::Encoding::decodeSymbols(const SymbolTable::Storage ar
   return symbol_vec;
 }
 
-void SymbolTableImpl::Encoding::decodeTokens(
-    const SymbolTable::Storage array, size_t size,
-    const std::function<void(Symbol)>& symbol_token_fn,
-    const std::function<void(absl::string_view)>& string_view_token_fn) {
-  while (size > 0) {
-    if (*array == LiteralStringIndicator) {
+// Iterator for decoding StatName tokens. This is not an STL-compatible
+// iterator; it is used only locally. Using this rather than decodeSymbols makes
+// it easier to early-exit out of a comparison once we know the answer. For
+// example, if you are ordering "a.b.c.d" against "a.x.y.z" you know the order
+// when you get to the "b" vs "x" and do not need to decode "c" vs "y".
+//
+// This is also helpful for prefix-matching.
+//
+// The array-version of decodeSymbols is still a good approach for converting a
+// StatName to a string as the intermediate array of string_view is needed to
+// allocate the right size for the joined string.
+class SymbolTableImpl::Encoding::TokenIter {
+public:
+  // The type of token reached.
+  enum class Type { StringView, Symbol, End };
+
+  TokenIter(const SymbolTable::Storage array, size_t size) : array_(array), size_(size) {}
```

Comment:
Explicitly delete default constructor to prevent having uninitialised array_ and size_?

Reply:
Aah okay. I looked for the relevant rule on cppreference but could not find it.